client/lobby/RandomMapTab.h

```diff
@@ -23,42 +25,80 @@ VCMI_LIB_NAMESPACE_END
 class CToggleButton;
 class CLabel;
 class CLabelGroup;
+class CSlider;
+class CPicture;
 
-class RandomMapTab : public CIntObject
+class RandomMapTab : public InterfaceObjectConfigurable
 {
 public:
 	RandomMapTab();
 
 	void updateMapInfoByHost();
 	void setMapGenOptions(std::shared_ptr<CMapGenOptions> opts);
+	void setTemplate(const CRmgTemplate *);
+	CMapGenOptions & obtainMapGenOptions() {return *mapGenOptions;}
 
 	CFunctionList<void(std::shared_ptr<CMapInfo>, std::shared_ptr<CMapGenOptions>)> mapInfoChanged;
 
 private:
-	void addButtonsWithRandToGroup(CToggleGroup * group, const std::vector<std::string> & defs, int startIndex, int endIndex, int btnWidth, int helpStartIndex, int helpRandIndex, int randIndex = -1, bool animIdfromBtnId = true) const;
-	void addButtonsToGroup(CToggleGroup * group, const std::vector<std::string> & defs, int startIndex, int endIndex, int btnWidth, int helpStartIndex, bool animIdfromBtnId = true) const;
-	void deactivateButtonsFrom(CToggleGroup * group, int startId);
-	void validatePlayersCnt(int playersCnt);
-	void validateCompOnlyPlayersCnt(int compOnlyPlayersCnt);
+	void deactivateButtonsFrom(CToggleGroup & group, const std::set<int> & allowed);
 	std::vector<int> getPossibleMapSizes();
 
-
-	std::shared_ptr<CPicture> background;
-	std::shared_ptr<CLabel> labelHeadlineBig;
-	std::shared_ptr<CLabel> labelHeadlineSmall;
-
-	std::shared_ptr<CLabel> labelMapSize;
-	std::shared_ptr<CToggleGroup> groupMapSize;
-	std::shared_ptr<CToggleButton> buttonTwoLevels;
-
-	std::shared_ptr<CLabelGroup> labelGroupForOptions;
-	std::shared_ptr<CToggleGroup> groupMaxPlayers;
-	std::shared_ptr<CToggleGroup> groupMaxTeams;
-	std::shared_ptr<CToggleGroup> groupCompOnlyPlayers;
-	std::shared_ptr<CToggleGroup> groupCompOnlyTeams;
-	std::shared_ptr<CToggleGroup> groupWaterContent;
-	std::shared_ptr<CToggleGroup> groupMonsterStrength;
-	std::shared_ptr<CButton> buttonShowRandomMaps;
 	std::shared_ptr<CMapGenOptions> mapGenOptions;
 	std::shared_ptr<CMapInfo> mapInfo;
+	
+	//options allowed - need to store as impact each other
+	std::set<int> playerCountAllowed, playerTeamsAllowed, compCountAllowed, compTeamsAllowed;
+};
+
+class TemplatesDropBox : public InterfaceObjectConfigurable
```

Comment:
Hmm... Kind of makes sense to have such widget is generic form, so it can be reused in any other new UI, since H3 UI lacks such drop box.
Don't want to end up with bunch of different drop box, with each having its own slightly different functionality.

Not sure whether this can be done in reasonable amount of time, so feel free to ignore.

Reply:
Not sure if I can do this generic because I rely on pop-up window functionality when show drop-box. I thought about it, but not trying to capture all possible scenarios for now.